### PR TITLE
Dịch notebook chap 04

### PR DESCRIPTION
**Pull request by quangnhat185**

### CODE REVIEW

File: 04_training_linear_models.ipynb [hunk 1]
```diff
@@ -4,14 +4,14 @@
...
-    "**Chapter 4 – Training Linear Models**"
+    "**Chapter 4 – Huấn luyện Mô hình Tuyến tính**"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "_This notebook contains all the sample code and solutions to the exercises in chapter 4._"
+    "_Notebook này chứa mã lập trình cũng như lời giải các bài tập tại chương 4._"
```
Comment (ngcthuong) [inline 04_training_linear_models.ipynb:14]:
```suggestion
    "_Notebook này chứa đoạn mẫu của mã lập trình cũng như lời giải các bài tập tại chương 4._"
```

File: 04_training_linear_models.ipynb [hunk 2]
```diff
@@ -32,14 +32,14 @@
...
-    "# Setup"
+    "# Cài đặt"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "First, let's import a few common modules, ensure MatplotLib plots figures inline and prepare a function to save the figures. We also check that Python 3.5 or later is installed (although Python 2.x may work, it is deprecated so we strongly recommend you use Python 3 instead), as well as Scikit-Learn ≥0.20."
+    "Đầu tiên, hạy nạp các mô-đun phổ biến, đảm bảo rằng Matplotlib sẽ vẽ các đồ thị ngay trong notebook, và thiết kế một hàm để lưu hình ảnh. Chúng ta cũng kiểm tra xem liệu Python 3.5 trở lên đã được cài đặt (Python 2.x mặc dù vẫn có thể hoạt động nhưng đã ít được sử dụng, do đó chúng tôi khuyên bạn nên sử dụng Python 3), cũng như Scikit-Learn ≥0.20."
```
Comment (ngcthuong) [inline 04_training_linear_models.ipynb:42]:
```suggestion
    "Đầu tiên, ta cần nạp các mô-đun phổ biến, đảm bảo rằng Matplotlib vẽ các đồ thị ngay trong notebook, và viết một hàm lưu hình ảnh. Ta cần kiểm tra xem liệu Python 3.5 trở lên đã được cài đặt chưa (dù Python 2.x vẫn có thể hoạt động nhưng đã ít được sử dụng, do đó chúng tôi khuyến nghị dùng Python 3), cũng như Scikit-Learn ≥0.20."
```

File: 04_training_linear_models.ipynb [hunk 3]
```diff
@@ -222,7 +222,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "The figure in the book actually corresponds to the following code, with a legend and axis labels:"
+    "Hình trong cuốn sách thực ra tương ứng với đoạn mã sau, với chú giải và nhãn trục:"
```
Comment (ngcthuong) [inline 04_training_linear_models.ipynb:225]:
```suggestion
    "Hình minh họa trong cuốn sách tương ứng với đoạn mã sau, với chú giải và tên trục:"
```

File: 04_training_linear_models.ipynb [hunk 4]
```diff
@@ -1364,7 +1368,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Create the graph:"
+    "Tạo đồ thị:"
```
Comment (ngcthuong) [inline 04_training_linear_models.ipynb:1371]:
```suggestion
    "Vẽ đồ thị:"
```

File: 04_training_linear_models.ipynb [hunk 5]
```diff
@@ -2144,7 +2148,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "And let's set the random seed so the output of this exercise solution is reproducible:"
+    "Và hãy đặt random seed cố định để có thể tái tạo lại đầu ra của lời giải này:"
```
Comment (ngcthuong) [inline 04_training_linear_models.ipynb:2151]:
```suggestion
    "Và hãy đặt random seed cố định để có thể tái tạo lại đầu ra của lời giải này:"
```
Sao mình không dịch từ random seed nhỉ?

File: 04_training_linear_models.ipynb [hunk 6]
```diff
@@ -2191,7 +2195,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "The targets are currently class indices (0, 1 or 2), but we need target class probabilities to train the Softmax Regression model. Each instance will have target class probabilities equal to 0.0 for all classes except for the target class which will have a probability of 1.0 (in other words, the vector of class probabilities for ay given instance is a one-hot vector). Let's write a small function to convert the vector of class indices into a matrix containing a one-hot vector for each instance:"
+    "Các nhãn mục tiêu hiện tại là chỉ số lớp (0, 1 hoặc 2), tuy nhiên chúng ta cần các nhãn mục tiêu xác suất để huấn luyện mô hình Hồi quy Softmax. Mỗi mẫu sẽ có xác suất các lớp mục tiêu không phải nhãn thực bằng 0.0, và lớp mục tiêu với nhãn thực sẽ có xác suất bằng 1.0 (nói cách khác, vector xác suất các lớp của một mẫu bất kỳ là một vector one-hot). Hãy viết một hàm đơn giản để chuyển đổi vector chỉ số lớp thành một ma trận chứa một vector one-hot cho mỗi mẫu:"
```
Comment (ngcthuong) [inline 04_training_linear_models.ipynb:2198]:
```suggestion
    "Các nhãn mục tiêu hiện tại là chỉ số lớp (0, 1, hoặc 2), tuy nhiên chúng ta cần các nhãn mục tiêu xác suất để huấn luyện mô hình Hồi quy Softmax. Mỗi mẫu sẽ có xác suất các lớp mục tiêu không phải nhãn thực bằng 0.0, và lớp mục tiêu với nhãn thực sẽ có xác suất bằng 1.0 (nói cách khác, vector xác suất các lớp của một mẫu bất kỳ là một vector one-hot). Hãy viết một hàm đơn giản để chuyển đổi vector chỉ số lớp thành một ma trận chứa một vector one-hot cho mỗi mẫu:"
```

File: 04_training_linear_models.ipynb [hunk 7]
```diff
@@ -2717,7 +2722,9 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Our perfect model turns out to have slight imperfections. This variability is likely due to the very small size of the dataset: depending on how you sample the training set, validation set and the test set, you can get quite different results. Try changing the random seed and running the code again a few times, you will see that the results will vary."
+    "Our perfect model turns out to have slight imperfections. This variability is likely due to the very small size of the dataset: depending on how you sample the training set, validation set and the test set, you can get quite different results. Try changing the random seed and running the code again a few times, you will see that the results will vary.\n",
```
Comment (ngcthuong) [inline 04_training_linear_models.ipynb:2725]:
```suggestion
```

File: 04_training_linear_models.ipynb (hunk 5 above)
Comment (quangnhat185) [reply 04_training_linear_models.ipynb:2151]: Hiện tại chưa có cách dịch chính thức trong glossary, nhưng theo em nên giữ nguyên để đồng bộ với code

File: 04_training_linear_models.ipynb [hunk 8]
```diff
@@ -32,14 +32,14 @@
...
-    "# Setup"
+    "# Cài đặt"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "First, let's import a few common modules, ensure MatplotLib plots figures inline and prepare a function to save the figures. We also check that Python 3.5 or later is installed (although Python 2.x may work, it is deprecated so we strongly recommend you use Python 3 instead), as well as Scikit-Learn ≥0.20."
+    "Đầu tiên, ta cần nạp các mô-đun phổ biến, đảm bảo rằng Matplotlib vẽ các đồ thị ngay trong notebook, và viết một hàm lưu hình ảnh. Ta cần kiểm tra xem liệu Python 3.5 trở lên đã được cài đặt chưa (dù Python 2.x vẫn có thể hoạt động nhưng đã ít được sử dụng, do đó chúng tôi khuyến nghị dùng Python 3), cũng như Scikit-Learn ≥0.20."
```
Comment (lkhphuc) [inline 04_training_linear_models.ipynb:42]:
```suggestion
    "Đầu tiên, ta cần nạp các mô-đun phổ biến, đảm bảo rằng Matplotlib vẽ các đồ thị ngay trong notebook, và viết một hàm lưu hình ảnh. Ta cần kiểm tra xem liệu Python 3.5 trở lên đã được cài đặt chưa (dù Python 2.x vẫn có thể hoạt động nhưng đã ít được sử dụng, do đó chúng tôi khuyến nghị dùng Python 3), và phiên bản Scikit-Learn ≥0.20."
```

File: 04_training_linear_models.ipynb [hunk 9]
```diff
@@ -222,7 +222,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "The figure in the book actually corresponds to the following code, with a legend and axis labels:"
+    "Hình minh họa trong cuốn sách tương ứng với đoạn mã sau, với chú giải và tên trục:"
```
Comment (lkhphuc) [inline 04_training_linear_models.ipynb:225]:
```suggestion
    "Hình minh họa trong cuốn sách tương ứng với đoạn mã sau, với chú thích và tên trục:"
```

File: 04_training_linear_models.ipynb [hunk 10]
```diff
@@ -339,7 +339,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "This function computes $\\mathbf{X}^+\\mathbf{y}$, where $\\mathbf{X}^{+}$ is the _pseudoinverse_ of $\\mathbf{X}$ (specifically the Moore-Penrose inverse). You can use `np.linalg.pinv()` to compute the pseudoinverse directly:"
+    "Hàm này tính $\\mathbf{X}^+\\mathbf{y}$, trong đó $\\mathbf{X}^{+}$ là _giả nghịch đảo_ của $\\mathbf{X}$ (cụ thể là nghịch đảo Moore-Penrose). Bạn có thể sử dụng `np.linalg.pinv()` để tính giả nghịch đảo trực tiếp:"
```
Comment (lkhphuc) [inline 04_training_linear_models.ipynb:342]:
```suggestion
    "Hàm này tính $\\mathbf{X}^+\\mathbf{y}$, trong đó $\\mathbf{X}^{+}$ là _giả nghịch đảo_ của $\\mathbf{X}$ (cụ thể là nghịch đảo Moore-Penrose). Bạn có thể sử dụng `np.linalg.pinv()` để trực tiếp tính giả nghịch đảo:"
```

File: 04_training_linear_models.ipynb [hunk 11]
```diff
@@ -1187,7 +1191,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "**Note**: to be future-proof, we set `max_iter=1000` and `tol=1e-3` because these will be the default values in Scikit-Learn 0.21."
+    "**Lưu ý**: để cập nhật với phiên bản hiện tại, chúng ta đặt `max_iter=1000` và `tol=1e-3` bởi chúng là các giá trị mặc định trong Scikit-Learn 0.21"
```
Comment (lkhphuc) [inline 04_training_linear_models.ipynb:1194]:
```suggestion
    "**Lưu ý**: để thống nhất với phiên bản trong tương lai, chúng ta đặt `max_iter=1000` và `tol=1e-3` bởi chúng là các giá trị mặc định trong Scikit-Learn 0.21"
```

File: 04_training_linear_models.ipynb [hunk 12]
```diff
@@ -1737,7 +1742,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "**Note**: To be future-proof we set `solver=\"lbfgs\"` since this will be the default value in Scikit-Learn 0.22."
+    "**Lưu ý**: Để cập nhật với phiên bản hiện tại, chúng ta đặt `solver=\"lbfgs\"` bởi đây là giá trị mặc định trong Scikit-Learn 0.22."
```
Comment (lkhphuc) [inline 04_training_linear_models.ipynb:1745]:
```suggestion
    "**Lưu ý**: Để thống nhất với phiên bản trong tương lai, chúng ta đặt `solver=\"lbfgs\"` bởi đây là giá trị mặc định trong Scikit-Learn 0.22."
```

File: 04_training_linear_models.ipynb [hunk 13]
```diff
@@ -2160,7 +2164,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "The easiest option to split the dataset into a training set, a validation set and a test set would be to use Scikit-Learn's `train_test_split()` function, but the point of this exercise is to try understand the algorithms by implementing them manually. So here is one possible implementation:"
+    "Tùy chọn dễ dàng nhất để chia tập dữ liệu thành một tập huấn luyện, một tập kiểm định, và một tập kiểm tra là sử dụng hàm `train_test_split()` của Scikit-Learn, tuy nhiên mục đích của bài tập này là cố gắng hiểu được thuật toán bằng cách triển khai chúng theo cách thủ công. Do đó, dưới đây là một cách triển khai có thể thực hiện:"
```
Comment (lkhphuc) [inline 04_training_linear_models.ipynb:2167]:
```suggestion
    "Tùy chọn dễ dàng nhất để chia tập dữ liệu thành một tập huấn luyện, một tập kiểm định, và một tập kiểm tra là sử dụng hàm `train_test_split()` của Scikit-Learn, tuy nhiên mục đích của bài tập này là cố gắng hiểu được thuật toán bằng cách triển khai chúng theo cách thủ công. Do đó, dưới đây là một cách thực hiện điều này:"
```

File: 04_training_linear_models.ipynb [hunk 14]
```diff
@@ -2321,22 +2325,23 @@
...
-    "\n",
-    "So the equations we will need are the cost function:\n",
+    "Bây giờ sẽ đến phần khó nhằn nhất: huấn luyện! Về mặt lý thuyết, nó tương đối đơn giản: chỉ việc chuyển đổi các phương trình toán học sang mã lập trình Python. Tuy nhiên trong thực tế, điều này khá là phức tạp: cụ thể là thứ tự các thuật ngữ và chỉ số rất dễ bị nhẫm lẫn. Thậm chí bạn có thể viết được một đoạn mã có vẻ như hoạt động tốt, nhưng thực ra lại không tính toán đúng như những gì ta muốn. Khi không chắc chắn, bạn nên viết ra hình dạng của từng thuật ngữ trong phương trình và đảm bảo rằng chúng khớp với các thuật ngữ tương ứng trong mã lập trình của bạn. Điều này giúp đánh giá từng thuật ngữ một cách độc lập cũng như in chúng ra. Tin tốt là bạn sẽ không phải làm điều này mỗi lần bởi tất cả đều được hỗ trợ đầy đủ bởi Scikit-Learn, tuy nhiên nó sẽ giúp bạn hiểu được những gì đang diễn ra ở phía sau.\n",
```
Comment (lkhphuc) [inline 04_training_linear_models.ipynb:2329]:
```suggestion
    "Bây giờ sẽ đến phần khó nhằn nhất: huấn luyện! Về mặt lý thuyết, nó tương đối đơn giản: chỉ việc chuyển đổi các phương trình toán học sang mã lập trình Python. Tuy nhiên trong thực tế, điều này khá là phức tạp: cụ thể là thứ tự các thuật ngữ và chỉ số rất dễ bị nhẫm lẫn. Thậm chí bạn có thể viết được một đoạn mã có vẻ như hoạt động tốt, nhưng thực ra lại không tính toán đúng như những gì ta muốn. Khi không chắc chắn, bạn nên viết ra hình dạng của từng thuật ngữ trong phương trình và đảm bảo rằng chúng khớp với các thuật ngữ tương ứng trong mã lập trình của bạn. Điều này giúp đánh giá từng thuật ngữ một cách độc lập cũng như in chúng ra. Tin tốt là bạn sẽ không phải làm điều này mỗi lần bởi tất cả đều được hỗ trợ đầy đủ bởi Scikit-Learn, tuy nhiên nó sẽ giúp bạn hiểu được những gì đang diễn ra ở bên dưới.\n",
```

File: 04_training_linear_models.ipynb [hunk 15]
```diff
@@ -2445,7 +2450,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Well, this model looks pretty good. For the sake of the exercise, let's add a bit of $\\ell_2$ regularization. The following training code is similar to the one above, but the loss now has an additional $\\ell_2$ penalty, and the gradients have the proper additional term (note that we don't regularize the first element of `Theta` since this corresponds to the bias term). Also, let's try increasing the learning rate `eta`."
+    "Mô hình này trông khá tốt. Theo yêu cầu của bài tập, hãy thêm vào một ít điều chuẩn $\\ell_2$. Đoạn mã huấn luyện dưới đây tương tự như đoạn trên, tuy nhiên giá trị mất mát nay được cộng thêm giá trị phạt $\\ell_2$, và gradient có thêm thuật ngữ cộng vào thích hợp (chú ý rằng chúng ta không điều chuẩn phần tử đầu tiên của `Theta` vì nó tương ứng với thuật ngữ thiên kiến). Ngoài ra, hãy thử tăng tốc độ học `eta`."
```
Comment (lkhphuc) [inline 04_training_linear_models.ipynb:2453]:
```suggestion
    "Mô hình này trông khá tốt. Theo yêu cầu của bài tập, hãy thêm vào một ít điều chuẩn $\\ell_2$. Đoạn mã huấn luyện dưới đây tương tự như đoạn trên, tuy nhiên giá trị mất mát nay được cộng thêm giá trị phạt $\\ell_2$, và gradient được cộng thêm vào một đại lượng thích hợp (chú ý rằng chúng ta không điều chuẩn phần tử đầu tiên của `Theta` vì nó tương ứng với đại lượng thiên kiến). Ngoài ra, hãy thử tăng tốc độ học `eta`."
```

File: 04_training_linear_models.ipynb [hunk 16]
```diff
@@ -2529,14 +2534,14 @@
...
-    "Cool, perfect accuracy! We probably just got lucky with this validation set, but still, it's pleasant."
+    "Tuyệt, chính xác tuyệt đối! Chúng ta có thể chỉ gặp may trên tập kiểm định này, tuy nhiên, điều này vẫn khá là tích cực "
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Now let's add early stopping. For this we just need to measure the loss on the validation set at every iteration and stop when the error starts growing."
+    "Bây giờ hãy thêm vào dừng sớm. Để làm được điều này, chúng ta chỉ cần đo giá trị mất mát trên tập kiểm định tại mỗi vòng lặp và dừng lại khi lỗi bắt đầu tăng. "
```
Comment (lkhphuc) [inline 04_training_linear_models.ipynb:2544]:
```suggestion
    "Bây giờ hãy thêm tính năng dừng sớm. Để làm được điều này, chúng ta chỉ cần đo giá trị mất mát trên tập kiểm định tại mỗi vòng lặp và dừng lại khi lỗi bắt đầu tăng. "
```

**Review by lkhphuc — APPROVED** [review]
Vài comment cuối.

**Review by duythanhvn — APPROVED** [review]
